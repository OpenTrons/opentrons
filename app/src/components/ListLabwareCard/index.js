// @flow
import * as React from 'react'
import { Card } from '@opentrons/components'

// TODO(mc, 2019-10-17): i18n
const LIST_LABWARE_CARD_TITLE = 'Custom Labware Listing'

function ListLabwareCard() {
  return <Card title={LIST_LABWARE_CARD_TITLE}></Card>
}

export default ListLabwareCard
