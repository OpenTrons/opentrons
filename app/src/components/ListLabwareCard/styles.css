@import '@opentrons/components';

.card {
  line-height: 1.5;
  padding-bottom: var(--spacing-2);
}

.list_column_titles {
  @apply --font-body-1-dark;

  padding: var(--spacing-3) var(--spacing-3) var(--spacing-2);
  text-transform: uppercase;
}

.item {
  @apply --font-body-2-dark;

  border-top: var(--bd-light);
  padding: var(--spacing-3);
}

.item_category_column,
.item_name_column,
.item_date_column {
  display: inline-block;
  vertical-align: top;
}

.item_category_column {
  text-align: right;
  width: 6rem;
}

.item_name_column {
  width: calc(100% - 14rem);
  padding-left: var(--spacing-3);
}

.item_date_column {
  width: 8rem;
}

.item_display_name,
.item_load_name {
  @apply --truncate;
}

.item_display_name {
  font-weight: var(--fw-semibold);
}
