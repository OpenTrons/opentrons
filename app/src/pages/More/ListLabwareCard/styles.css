@import '@opentrons/components';

.labware_list_wrapper {
  @apply --font-body-1-dark;

  padding: 1rem 0;
}

.unstyled_list {
  list-style: none;
}

.message_wrapper {
  padding: 0 1rem 0.5rem;
}

.warning {
  display: flex;
  align-items: flex-start;
  color: var(--c-warning);
}

.warning_icon {
  flex: none;
  width: 1.5rem;
  margin-top: 0.25rem;
  margin-right: 1rem;
}

.error_message,
.item_warning_wrapper {
  margin-top: 0.75rem;
}

.italic {
  font-style: italic;
}

.list_column_titles {
  display: flex;
  padding: 1rem;
  font-size: var(--fs-caption);
  text-transform: uppercase;
}

.item {
  display: flex;
  align-items: flex-start;
  border-top: var(--bd-light);
  padding: 1rem;
}

.invalid {
  color: var(--c-med-gray);
}

.item_category_column,
.item_date_column {
  flex: none;
  width: 5rem;
}

.item_date_column {
  text-align: center;
}

.item_name_column {
  width: 100%;
  padding-left: 1rem;
  padding-right: 1rem;
}

.item_primary_name {
  font-weight: var(--fw-semibold);
}

.item_titled_name {
  display: flex;
  align-items: baseline;
  margin-top: 0.25rem;
}

.titled_name_title {
  flex: none;
  width: 4rem;
  font-size: var(--fs-caption);
  text-transform: uppercase;
}
