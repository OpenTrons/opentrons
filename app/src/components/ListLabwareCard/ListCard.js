// @flow
import * as React from 'react'
import { Card } from '@opentrons/components'

import styles from './styles.css'

// TODO(mc, 2019-10-22): i18n
const LIST_LABWARE_CARD_TITLE = 'Custom Labware Listing'
const NAME = 'Name'
const DATE_ADDED = 'Date Added'

export type ListCardProps = {|
  children?: React.Node,
|}

export default function ListCard(props: ListCardProps) {
  return (
    <Card title={LIST_LABWARE_CARD_TITLE} className={styles.card}>
      <div className={styles.list_column_titles}>
        <p className={styles.item_category_column} />
        <p className={styles.item_name_column}>{NAME}</p>
        <p className={styles.item_date_column}>{DATE_ADDED}</p>
      </div>
      <ul>{props.children}</ul>
    </Card>
  )
}
