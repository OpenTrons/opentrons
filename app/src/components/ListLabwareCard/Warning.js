// @flow
import * as React from 'react'

import { Icon } from '@opentrons/components'
import styles from './styles.css'

export type WarningProps = {|
  children: React.Node,
|}

export function Warning(props: WarningProps): React.Node {
  const { children } = props

  return (
    <div className={styles.warning}>
      <Icon className={styles.warning_icon} name="alert-circle" />
      <div>{children}</div>
    </div>
  )
}
