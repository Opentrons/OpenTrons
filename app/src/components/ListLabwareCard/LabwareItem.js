// @flow
import * as React from 'react'
import cx from 'classnames'
import startCase from 'lodash/startCase'
import { basename } from 'path'
import { format } from 'date-fns'

import {
  INVALID_LABWARE_FILE,
  DUPLICATE_LABWARE_FILE,
  OPENTRONS_LABWARE_FILE,
} from '../../redux/custom-labware'

import { Warning } from './Warning'
import styles from './styles.css'

import type { CheckedLabwareFile } from '../../redux/custom-labware/types'

// TODO(mc, 2019-11-21): i18n
const NA = 'N/A'
const FILE_NAME = 'File Name'
const API_NAME = 'API Name'

const THIS_FILE_IS_NOT_A_VALID_DEFINITION =
  'This file is not a valid Opentrons labware definition or cannot be read. Please check the file.'

const THIS_FILE_CONFLICTS_WITH_ANOTHER_DEFINITION =
  'This file conflicts with another labware definition that is already in this folder. Please check the API name.'

const THIS_FILE_CONFLICTS_WITH_AN_OPENTRONS_DEFINITION =
  'This file conflicts with an Opentrons standard labware definition. If you are trying to create a definition based on an Opentrons definition, please contact support'

export type LabwareItemProps = {|
  file: CheckedLabwareFile,
|}

export function LabwareItem(props: LabwareItemProps): React.Node {
  const { file } = props
  const { type, filename, modified, definition = null, errors = null } = file
  const apiName = definition?.parameters.loadName || NA
  const displayName = definition?.metadata.displayName || NA
  const displayCategory = definition
    ? startCase(definition.metadata.displayCategory)
    : NA

  let warning = null

  if (type === INVALID_LABWARE_FILE) {
    warning = THIS_FILE_IS_NOT_A_VALID_DEFINITION
  } else if (type === DUPLICATE_LABWARE_FILE) {
    warning = THIS_FILE_CONFLICTS_WITH_ANOTHER_DEFINITION
  } else if (type === OPENTRONS_LABWARE_FILE) {
    warning = THIS_FILE_CONFLICTS_WITH_AN_OPENTRONS_DEFINITION
  }

  return (
    <li className={cx(styles.item, { [styles.invalid]: warning })}>
      <p className={styles.item_category_column}>{displayCategory}</p>
      <div className={styles.item_name_column}>
        <p className={styles.item_primary_name}>{displayName}</p>
        <p className={styles.item_titled_name}>
          <span className={styles.titled_name_title}>{FILE_NAME}</span>
          <span>{basename(filename)}</span>
        </p>
        <p className={styles.item_titled_name}>
          <span className={styles.titled_name_title}>{API_NAME}</span>
          <span>{apiName}</span>
        </p>
        {warning !== null && (
          <div className={styles.item_warning_wrapper}>
            <Warning>
              <p>{warning}</p>
              {errors &&
                errors.map(({ dataPath, message }) => (
                  <p>
                    {dataPath} {message}
                  </p>
                ))}
            </Warning>
          </div>
        )}
      </div>
      <p className={styles.item_date_column}>
        {format(new Date(modified), 'yyyy-MM-dd')}
      </p>
    </li>
  )
}
