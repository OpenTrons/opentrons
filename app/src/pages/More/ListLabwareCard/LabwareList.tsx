import * as React from 'react'

import { LabwareItem } from './LabwareItem'
import { Warning } from './Warning'
import styles from './styles.css'

import type { CheckedLabwareFile } from '../../../redux/custom-labware/types'

// TODO(mc, 2019-10-22): i18n
const CATEGORY = 'Category'
const LABWARE = 'Labware'
const DATE_ADDED = 'Date Added'

const UNABLE_TO_READ_CUSTOM_LABWARE =
  'Unable to read Custom Labware source folder.'

const PLEASE_ENSURE_CORRECT_FOLDER =
  'Please ensure you are able to access the folder selected in Labware Management above.'

const NO_LABWARE_DEFINITIONS_FOUND = 'No labware definitions found'

const PLEASE_ADD_CUSTOM_LABWARE =
  'Please add custom labware via the Labware Management options above'

export interface LabwareListProps {
  labware: CheckedLabwareFile[]
  errorMessage: string | null
}

export function LabwareList(props: LabwareListProps): JSX.Element {
  const { labware, errorMessage } = props
  let children = null

  if (errorMessage !== null) {
    children = (
      <div className={styles.message_wrapper}>
        <Warning>
          <p>
            {UNABLE_TO_READ_CUSTOM_LABWARE}
            <br />
            {PLEASE_ENSURE_CORRECT_FOLDER}
          </p>
          <p className={styles.error_message}>{errorMessage}</p>
        </Warning>
      </div>
    )
  } else if (labware.length === 0) {
    children = (
      <div className={styles.message_wrapper}>
        <p className={styles.italic}>
          {NO_LABWARE_DEFINITIONS_FOUND}
          <br />
          {PLEASE_ADD_CUSTOM_LABWARE}
        </p>
      </div>
    )
  } else {
    children = (
      <>
        <div className={styles.list_column_titles}>
          <p className={styles.item_category_column}>{CATEGORY}</p>
          <p className={styles.item_name_column}>{LABWARE}</p>
          <p className={styles.item_date_column}>{DATE_ADDED}</p>
        </div>
        <ul className={styles.unstyled_list}>
          {labware.map(lwFile => (
            <LabwareItem key={lwFile.filename} file={lwFile} />
          ))}
        </ul>
      </>
    )
  }

  return <div className={styles.labware_list_wrapper}>{children}</div>
}
