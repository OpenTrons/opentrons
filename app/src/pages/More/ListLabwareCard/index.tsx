import * as React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Card, useInterval } from '@opentrons/components'

import { LabwareList } from './LabwareList'
import {
  fetchCustomLabware,
  getCustomLabware,
  getListLabwareErrorMessage,
} from '../../../redux/custom-labware'

import type { Dispatch } from '../../../redux/types'

const LABWARE_REFRESH_INTERVAL_MS = 5000

// TODO(mc, 2019-11-21): i18n
const CUSTOM_LABWARE_LIST = 'Custom Labware List'

export function ListLabwareCard(): JSX.Element {
  const dispatch = useDispatch<Dispatch>()
  const labware = useSelector(getCustomLabware)
  const listLabwareError = useSelector(getListLabwareErrorMessage)
  const fetchLabware = React.useCallback(() => dispatch(fetchCustomLabware()), [
    dispatch,
  ])

  useInterval(fetchLabware, LABWARE_REFRESH_INTERVAL_MS, true)

  return (
    <Card title={CUSTOM_LABWARE_LIST}>
      <LabwareList labware={labware} errorMessage={listLabwareError} />
    </Card>
  )
}
