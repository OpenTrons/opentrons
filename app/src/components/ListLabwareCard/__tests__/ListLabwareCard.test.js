// @flow
import * as React from 'react'
import { shallow } from 'enzyme'

import ListLabwareCard from '..'

describe('ListLabwareCard', () => {
  test('component renders', () => {
    const tree = shallow(<ListLabwareCard />)

    expect(tree).toMatchSnapshot()
  })
})
