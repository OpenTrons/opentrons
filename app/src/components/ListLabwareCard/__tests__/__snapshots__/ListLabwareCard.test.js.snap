// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListLabwareCard component renders 1`] = `
<Card
  title="Custom Labware Listing"
/>
`;
